src/test/java/com/snowplowanalytics/redash/RedashClientDataSourceTest.java

```diff
@@ -0,0 +1,246 @@
+        int id = redashClient.createDataSource(rds);
+        redashClient.addDataSourceToGroup(id, defaultGroup.getId() + 1);
+    }
+
+    @Test(expected = IOException.class)
+    public void addDataSourceToGroupWithWrongClientTest() throws IOException {
+        int id = redashClient.createDataSource(rds);
+        wrongClient.addDataSourceToGroup(id, defaultGroup.getId());
+    }
+
+    @Test
+    public void removeDataSourceFromGroupTest() throws IOException {
+        int id = redashClient.createDataSource(rds);
+        UserGroup groupFromDb = redashClient.getWithUsersAndDataSources(defaultGroup.getId());
+        Assert.assertTrue(groupFromDb.getDataSources().size() == 1);
+        Assert.assertTrue(redashClient.removeDataSourceFromGroup(id, defaultGroup.getId()));
+        groupFromDb = redashClient.getWithUsersAndDataSources(defaultGroup.getId());
+        Assert.assertTrue(groupFromDb.getDataSources().isEmpty());
+        Assert.assertFalse(redashClient.removeDataSourceFromGroup(id, defaultGroup.getId()));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void removeNonExistingDataSourceFromGroupTest() throws IOException {
+        int id = redashClient.createDataSource(rds);
+        redashClient.removeDataSourceFromGroup(id + 1, defaultGroup.getId());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void removeDataSourceFromNonExistingGroupTest() throws IOException {
+        int id = redashClient.createDataSource(rds);
+        redashClient.removeDataSourceFromGroup(id, defaultGroup.getId() + 1);
+    }
+
+    @Test(expected = IOException.class)
+    public void removeDataSourceFromGroupWithWrongClientTest() throws IOException {
+        int id = redashClient.createDataSource(rds);
+        wrongClient.removeDataSourceFromGroup(id, defaultGroup.getId());
+    }
+
+    public static boolean dataSourceMatcher(DataSource first, DataSource second) {
+        if (first.getName() == null || first.getType() == null) {
+            return false;
+        }
+        return first.getName().equals(second.getName()) &&
+                first.getType().equals(second.getType()) &&
+                first.getHost().equals(second.getHost()) &&
+                first.getPort() == second.getPort() &&
+                first.getDbName().equals(second.getDbName()) &&
+                first.getUser().equals(second.getUser());
+    }
+
+    public static boolean simpleDatasourceMatcher(DataSource dataSource) {
+        if (dataSource.getName() == null || dataSource.getType() == null) {
+            return false;
+        }
+        return dataSource.getName().equals(rds.getName()) && dataSource.getType().equals(rds.getType());
+    }
+
+    private void wipeDataSources() throws IOException {
+        List<DataSource> dataSources = redashClient.getDataSources();
+        if  (dataSources.isEmpty()) {
+            return;
+        }
+        for(DataSource ds:dataSources){
```

Comment:
Formatting: `for(DataSource ds:dataSources){` > `for (DataSource ds : dataSources) {`